### PR TITLE
docs: add notification section to customize and reference

### PR DESCRIPTION
**Pull request by rekt-hard**
:heart: Thank you for your contribution!

### Description

closes https://github.com/inveniosoftware/invenio-app-rdm/issues/2204

Adds notification documentation to the reference and customization sections.

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [ ] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [ ] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [ ] I've added relevant test cases.
- [ ] I've added relevant documentation.
- [ ] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/best-practices/i18n/) (for relevant code).
- [ ] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines (for relevant code).
- [ ] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines (for relevant code).
- [ ] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines (for relevant code).
- [ ] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [ ] I've NOT included third-party code (copy/pasted source code or new dependencies).

**Third-party code**

If you've added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect](https://github.com/orgs/inveniosoftware/teams/architects/members).

**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

### CODE REVIEW

File: docs/customize/notifications.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
+# Notifications
+
+Depending on your instance needs regarding notifications, you might want to customize the configuration values provided by the `invenio-notifications` module.
```
Comment (ppanero) [inline docs/customize/notifications.md:3]:
```suggestion
Depending on your instance needs regarding notifications, you might want to customize the default values provided by the `invenio-notifications` module.
```

File: docs/customize/notifications.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+<a href="{{ notification.context.get("request").get("receiver").get("links").get("self_html") + "/requests" }}" class="button">Check out all community requests</a>
+
+
+{% endblock %}
+```
+
+## NOTIFICATIONS_BACKENDS
+
+This config variable allows to specify the available backends.
+For instance, you can provide an implementation of your institute specific tool of communication and send notifications via this backend. For this, simply extend the `NotificationBackend` class and implement the `send` method.
```
Comment (ppanero) [inline docs/customize/notifications.md:91]: Mainly to use the same terminology than in the example below (i.e. institute vs institution)
```suggestion
For instance, you can provide an implementation for your institution's preferred communication tool and send notifications via this backend. For this, simply extend the `NotificationBackend` class and implement the `send` method.
```

File: docs/reference/notifications.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
+# Notifications
+
+_Introduced in InvenioRDM v12_
+
+The notification system in InvenioRDM is provided via the [`invenio-notifications`]() module.
```
Comment (ppanero) [inline docs/reference/notifications.md:5]: missing link in the `()`

File: docs/reference/notifications.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+```py
+@dataclass
+class Notification:
+
+    type: str
+    context: dict
+```
+
+The context attribute of the notification holds information relevant for further processing and will be expanded throughout the workflow.
+When dispatching this notification in the service, the context shall be as minimal as possible, to reduce passing huge chunks of data.
```
Comment (ppanero) [inline docs/reference/notifications.md:30]: > to reduce passing huge chunks of data.

Maybe add what are the implications in case that it was done (i.e. performance, etc.)

File: docs/reference/notifications.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+_EntityResolve_
+
+This is a discrete implementation of the context generator, using the registered entity resolvers of the `invenio-notifications` module to expand the context. It will take a key to do a dictionary look-up and store the resolved entity at the index of the key.
+
+```py
+class EntityResolve(ContextGenerator):
+    """Payload generator for a notification using the entity resolvers."""
+
+    def __init__(self, key):
+        """Ctor."""
```
Comment (ppanero) [inline docs/reference/notifications.md:78]:
```suggestion
        """Constructor."""
```

**Review by ppanero — COMMENTED** [review]
Massive work! Thanks a lot! I have two main comments:
- The cuzomize part talks about the notification backends and builders but does not specify what they are. While I do agree that might not be the place to do so, pointers to the respective reference sections would help on navigating it.
- The reference sections starts already mentioning backends and builders from the first paragraph. However, those concepts have not been introduced and makes is a bit hard to follow. The end result is that all the pieces are not put together until the end. (biased opinion now...) I am more a fan of gradually introducing the concepts (i.e. in the first paragraphs you can put a one line definition to people can understand what you are talking about, then it gets fully explained later on).

File: docs/reference/notifications.md (hunk 4 above)
Comment (rekt-hard) [reply docs/reference/notifications.md:35]: Added implications of passing a large context/notification

File: docs/reference/notifications.md [hunk 6]
```diff
@@ -0,0 +1,302 @@
+# Notifications
+
+_Introduced in InvenioRDM v12_
+
+The notification system in InvenioRDM is provided via the [`invenio-notifications`](https://github.com/inveniosoftware/invenio-notifications) module.
+This module provides customization possibilities for:
+
+- [Notification Backends](#backends): A notification backend is responsible for the actual sending of the notification to a recipient.
+- [Notification Builders](#): A notification builder specifies builders and filters in order to to construct the notification context, fetch recipients, filter recipients and choose which backends should be used to send the notification.
```
Comment (alejandromumo) [inline docs/reference/notifications.md:9]: empty reference `#`

File: docs/reference/notifications.md [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    def __call__(self, notification, recipients):
+        """Update required recipient information and add backend id."""
+        user = dict_lookup(notification.context, self.key)
+        recipients[user["id"]] = Recipient(data=user)
+        return recipients
+```
+
+#### RecipientFilter
+
+A recipient filter will get the fully expanded notification and all created recipients. The task of the filter is to filter recipients in place, which do not fulfill a certain requirement.c
```
Comment (alejandromumo) [inline docs/reference/notifications.md:137]: typo: "requirement.c"

File: docs/reference/notifications.md [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        return recipients
+```
+
+#### RecipientBackendGenerator
+
+A recipient backend generator will get the fully expanded notification, a single recipient and previously created backend ids. The task of this generator is to return the id of the backend, it wants to send the notification to. Additionally, if the backend depends on information, not yet available in the recipient, it can modify the recipient in place and provide this information.
+
+_UserEmailBackend_
+
+This is a discrete implementatio of the recipient backend generator. When called, it will add the id of the email backend to the backends parameter. Since all information needed for sending mails are already available in the recipient, it does not have to do any additonal work.
```
Comment (alejandromumo) [inline docs/reference/notifications.md:167]: typo: "discrete implementatio" (missing 'n')

File: docs/reference/notifications.md [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+        })
+        return resp
+```
+
+## Templates
+
+The Jinja templates provided shall include all parts of the notification that are subject to special formatting by a backend (e.g. subject, HTML/plaintext/markdown body, etc.) in separate Jinja blocks.
+Additional backends should provide their own templates, to be as specific as possible.
+
+This is an example notifitication
```
Comment (alejandromumo) [inline docs/reference/notifications.md:229]: typo: 'notifitication' (instead of 'notification')

**Review by alejandromumo — COMMENTED** [review]
Left minor questions

File: docs/reference/notifications.md (hunk 6 above)
Comment (Samk13) [inline docs/reference/notifications.md:9]:
```suggestion
- [Notification Builders](#): A notification builder specifies builders and filters in order to construct the notification context, fetch recipients, filter recipients and choose which backends should be used to send the notification.
```

**Comment by rekt-hard** [comment]
Notifications walk-through page

image

File: docs/features/features-walk-through/notifications.md [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+Notifications are sent at specific events (e.g. community invitation, community recod submission), to inform users of relevant actions.
+For instance customization possibilities, see [customize notifications](../../customize/notifications.md)
+
+## Notification Settings
+
+Available notification settings can be accessed via the account settings page (`<domain>/account/settings/notifications/`). Additionally, the main menu has an entry for easy access:
+
+![Notifications Setting Menu Entry](./img/notifications/menu.png "Notification Settings Menu Entry")
+
+For personal preferences, it is possible to completely enable or disable receiving notifications. The info on the right hand side will show the channel on which notifications will be received.
```
Comment (Samk13) [inline docs/features/features-walk-through/notifications.md:12]: // Suggestion: Include a config variable for users to set notifications on by default, overriding the current 'off' default setting.

File: docs/reference/notifications.md [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+The record "{{ notification.context.get("request").get("topic").get("metadata").get("title") }}" was submitted to your community {{ notification.context.get("request").get("receiver").get("metadata").get("title") }} by {{ notification.context.get("request").get("created_by").get("username")}}.
+
+[Review the request]({{ notification.context.get("request").get("links").get("self_html") }})
+{%- endblock md_body -%}
+```
+
+## Configuration Values
+
+Configuration values used in the `invenio-notifications` module can be overriden, in order to adapt instances to specific needs.
+
```
Comment (Samk13) [inline docs/reference/notifications.md:261]: Suggestion:
Email notifications lack port in links if it's not explicitly added in either config:
```python
SITE_UI_URL = "https://127.0.0.1:5000"
SITE_API_URL = "https://127.0.0.1:5000/api"
```
or cli (e.g. invenio-cli run --port 5000). The celery worker appears to struggle to read the current port value. Therefore, to me, it's a good idea if we document that specifying the port number is necessary for the proper functioning of notification links."

**Review by zzacharo — COMMENTED** [review]
Hey @rekt-hard, I will summarize here what we discussed IRL:

- Let's split the PR to 2 based on the audience, i.e [users](https://github.com/inveniosoftware/docs-invenio-rdm/issues/558) and [developers](https://github.com/inveniosoftware/invenio-app-rdm/issues/2204)
- I would take inspiration from the existing examples of [custom fields](https://inveniordm.docs.cern.ch/customize/metadata/custom_fields/records/#records-custom-fields) and [administration panel](https://inveniordm.docs.cern.ch/develop/topics/administration_panel/) on how to expose the feature documentation based on a real example so users can relate and gather the information they want
- **User audience**: 
  - **Feature walkthrough**: High level overview of the notifications feature. A user should quickly understand what is this feature about, maybe with an example.
  - **Customize**: I think users cannot customize notifications without developing experience so I would not add a section there. If that's not the case then we can add it!
- **Developer audience**:
  - **Reference**: I would keep only the notification classes definitions with explanations of the properties and methods if it makes sense
  - **Architecture**: I would explain there the notifications feature from the design perspective. For example give the dictionary of the involved entities i.e receiver, backend etc. Inspiration could be the [Events handling](https://inveniordm.docs.cern.ch/develop/architecture/event_handling/).
  - **Develop/Topic guide**: I would explain there how you can build your own notification backend (if that's the case currently or omit otherwise)
  - **Develop/ How to**: Maybe add there how I can change the notification template or other configurable parts of the feature
 
We had a really nice tutorial in the past on general information on how to write documentation that personally I always go back and refresh my memory 😅  You can find it here if you think is useful for you https://inveniordm.docs.cern.ch/maintenance/documentation/.
